src/Java.Interop/Java.Interop/ManagedPeer.cs

```diff
@@ -119,87 +119,153 @@ internal static void Init ()
 			}
 		}
 
-		static Exception CreateMissingConstructorException (Type type, Type [] ptypes)
+		static Exception CreateMissingConstructorException (Type type, string signature)
 		{
 			var message = new StringBuilder ();
-			message.Append ("Unable to find constructor ");
+			message.Append ("Unable to find constructor for type `");
 			message.Append (type.FullName);
-			message.Append ("(");
-
-			if (ptypes.Length > 0) {
-				message.Append (ptypes [0].FullName);
-				for (int i = 1; i < ptypes.Length; ++i)
-					message.Append (", ").Append (ptypes [i].FullName);
-			}
-
-			message.Append (")");
-			message.Append (". Please provide the missing constructor.");
+			message.Append ("` with JNI signature `");
+			message.Append (signature);
+			message.Append ("`. Please provide the missing constructor.");
 
 			return new NotSupportedException (message.ToString (), CreateJniLocationException ());
 		}
 
-
 		static Exception CreateJniLocationException ()
 		{
 			using (var e = new JavaException ()) {
 				return new JniLocationException (e.ToString ());
 			}
 		}
 
-		static Type[] GetParameterTypes (string? signature)
+		static Dictionary<string, ConstructorInfo?> ConstructorCache    = new Dictionary<string, ConstructorInfo?> ();
+
+		static ConstructorInfo? GetConstructor (Type type, string jniTypeName, string signature)
 		{
-			if (string.IsNullOrEmpty (signature))
-				return Array.Empty<Type> ();
-			var typeNames   = signature!.Split (':');
-			var ptypes      = new Type [typeNames.Length];
-			for (int i = 0; i < typeNames.Length; i++)
-				ptypes [i] = Type.GetType (typeNames [i], throwOnError:true)!;
-			return ptypes;
+			var ctorCacheKey    = jniTypeName + "." + signature;
+			lock (ConstructorCache) {
+				if (ConstructorCache.TryGetValue (ctorCacheKey, out var ctor)) {
+					return ctor;
+				}
+			}
+
+			var candidateParameterTypes = GetConstructorCandidateParameterTypes (signature);
+			if (candidateParameterTypes.Length == 0) {
+				return CacheConstructor (ctorCacheKey, type.GetConstructor (Array.Empty<Type> ()));
+			}
+
+			var constructors    = new List<ConstructorInfo>(type.GetConstructors (BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance));
+
+			// Filter out wrong parameter count
+			for (int c = constructors.Count; c > 0; --c) {
+				if (constructors [c-1].GetParameters ().Length != candidateParameterTypes.Length) {
+					constructors.RemoveAt (c-1);
+				}
+			}
+
+			// Filter out mismatched types
+			for (int c = constructors.Count; c > 0; --c) {
+				var parameters  = constructors [c-1].GetParameters ();
+				for (int i = 0; i < parameters.Length; ++i) {
+					if (!candidateParameterTypes [i].Contains (parameters [i].ParameterType)) {
```

Comment:
As mentioned [above](https://github.com/xamarin/java.interop/pull/1168#discussion_r1408402443), I'm concerned about app startup time overheads, and I thus believe/assert that HashSet/Dictionary/etc. creation and population will take much more time than an array, and because this array is (1) small, and (2) never changed, init time *should* be minimal.  (But I should verify this.)

Reply:
Apparently `<details/>` doesn't work like i thought, so here's the execution + summary:

Results:

```sh
% dotnet run
Array Creation: 646ms
 Dict Creation: 1131ms
Array Lookup: 29ms
 Dict Lookup: 1131ms

% dotnet run -c Release
Array Creation: 389ms
 Dict Creation: 770ms
Array Lookup: 21ms
 Dict Lookup: 770ms
```

As expected, array lookup is faster -- half the time! -- and lookup is also much faster with arrays.

That said, this investigation did present a bug in `GetBuiltInTypeArraySignature()`: it shouldn't have been looking at `.ArrayTypes`, but instead `.PrimitiveType`, because the callsite in `GetTypeSignature(Type)` "unwraps" arrays, which kinda makes that particular codepath unusable, as in `GetBuiltInTypeArraySignature()` should just be removed entirely. :thinking:

Reply:
I think I was confused [above](https://github.com/xamarin/java.interop/pull/1168#discussion_r1408403858); this is from `ManagedPeer.Construct()`, which isn't part of app startup, and no static data is in the picture.

Despite being confused, I'm think I'm more or less right, though: each element of `candidateParameterTypes` will be a *small* list, likely no more than 5 elements.  I believe that the overhead of a HashSet vs. a List at such small sizes is not warranted.  Arrays are fast!